src/Form/matching.js

```diff
@@ -500,24 +503,133 @@ class Matching {
     }
 }
 
+/**
+ *  @returns {SupportedTypes}
+ */
+function getInputType (input) {
+    const attr = input.getAttribute(ATTR_INPUT_TYPE)
+    if (isValidSupportedType(attr)) {
+        return attr
+    }
+    return 'unknown'
+}
+
+/**
+ * Retrieves the main type
+ * @param {SupportedTypes} type
+ * @returns {SupportedMainTypes}
+ */
+function getMainTypeFromType (type) {
+    const mainType = type.split('.')[0]
+    switch (mainType) {
+    case 'credentials':
+    case 'creditCard':
+    case 'identities':
+        return mainType
+    }
+    return 'unknown'
+}
+
 /**
  * Retrieves the input main type
  * @param {HTMLInputElement} input
- * @returns {SupportedMainTypes | string}
+ * @returns {SupportedMainTypes}
  */
 const getInputMainType = (input) =>
-    input.getAttribute(ATTR_INPUT_TYPE)?.split('.')[0] ||
+    getMainTypeFromType(getInputType(input))
+
+/** @typedef {supportedSubtypes[number]} SupportedSubTypes */
+const supportedSubtypes = /** @type {const} */ ([
```

Comment:
With that we get the error:
```
src/Form/matching.js:569:49 - error TS1355: A 'const' assertions can only be applied to references to enum members, or string, number, boolean, array, or object literals.
```

I agree the duplication isn't ideal but it would have all been in the d.ts file but most of it was missing.

We could use:
/** @typedef {`${SupportedMainTypes}.${SupportedSubTypes}`} SupportedTypes */

Which would include types that are invalid, the alternative is what I've submitted in the latest which gives us the most safety but also feels fragile in other ways.